src/sparknacl-hkdf.adb

```diff
@@ -0,0 +1,60 @@
+with SPARKNaCl.MAC;
+
+package body SPARKNaCl.HKDF
+  with SPARK_Mode => On
+is
+
+   --------------------------------------------------------
+   --  SHA-256 HKDF-Extract per RFC 5869
+   --------------------------------------------------------
+
+   procedure Extract (PRK  :    out Hashing.Digest_256; -- output key
+                      IKM  : in     Byte_Seq; -- input keying material
+                      Salt : in     Byte_Seq) -- input salt (can be 0 bytes)
+   is
+   begin
+      MAC.HMAC_SHA_256 (PRK, IKM, Salt);
+   end Extract;
+
+   --------------------------------------------------------
+   --  SHA-256 HKDF-Expand per RFC 5869
+   --------------------------------------------------------
+
+   procedure Expand (OKM  :    out OKM_256;   -- output key material
+                     PRK  : in     Hashing.Digest_256; -- pseudo-random key
+                     Info : in     Byte_Seq)   -- optional context
+   is
+      type T_Arr is array (N32 range 1 .. (OKM'Length + 31) / 32) of Bytes_32;
+
+      T : T_Arr with Relaxed_Initialization;
+   begin
+      pragma Assert (OKM'Last <= T'Last * 32);
+
+      MAC.HMAC_SHA_256 (T (1), Info & 1, PRK);
+
+      for I in 2 .. T_Arr'Last loop
+         pragma Loop_Invariant (T (1 .. I - 1)'Initialized);
+         MAC.HMAC_SHA_256 (T (I), T (I - 1) & Info & Byte (I), PRK);
```

Comment:
Let me see what I can do - I think I can refactor it to just use a couple temp arrays "Ti" representing T(i - 1) and "Tj" representing T(i) and then keep track of the output index and assign to OKM as it goes through the first loop.

Reply:
I was able to make this change - it adds a bit of complexity since depending on the output length (OKM'Length) you can't always assign the whole T block, but it should save some stack space and be a bit more clear.

I'm taking another look at HMAC too, I was playing around with some of the TLS 1.3 HKDF functionality (implementation of the HKDF-Expand-Label function defined in https://www.rfc-editor.org/rfc/rfc8446.html) and have been getting some differing results than the test vectors in https://datatracker.ietf.org/doc/html/rfc8448). I need to find some more test vectors and put this through the wringer a bit.

Reply:
Did you commit this change?  Apologies for slowness over the last couple of weeks - been catching up with work stuff following the Easter break.

Reply:
No worries - I got a bit sidetracked when I ran into that issue resulting from my SHA-256 goof. I was able to rework this particular subprogram to use a single T element but haven’t yet proved it out. The commits are in my fork’s tlscipher branch but I need to do some cleanup, implement your other recommendations, and make sure everything proves before I re-submit a PR.

On some positive notes, I added many more test cases as well for the SHA-256, HMAC SHA-256 and HKDF functions which will be in the new PR. I’d eventually like to incorporate the full set of NIST test vectors and the Wycheproof suite.

I have been working on a little toy TLS client which makes use of the existing ec25519 and Poly1305 functions in SPARKNaCl and the new HKDF and ChaCha20 work I’ve been doing and seeing some good results when talking with OpenSSL’s s_server.

It’s doing the initial handshake key exchange and secret generation and derivation successfully (once I fixed my mistake in SHA-256), and now I’m learning the wonderful world of ASN.1/DER decoding… it’s been a good testbed for SPARKNaCl!

Reply:
OK... let me know when things are ready to review.  I have some experience with ASN.1/DER ... it's entertaining... I highly recommend the book "ASN.1 The Tutorial and Reference" by Douglas Steedman if you can find a copy...